### ISSUE TITLE
Is it possible to dynamically construct compute graph?

### ISSUE THREAD

**Issue by ifsheldon** [neuronika/neuronika#89]
It seems the laziness of the compute graph construction does not allow dynamic control flows that depend on data values?

For example, in PyTorch, we can do:
```python
def foo(x):
    if x > 5: # suppose x is a scalar tensor
        return x * 2
    else:
        return x 
``` 
And we can even have loops that depend on the values and/or shapes of tensors. Now it seems neuronika does not have such capabilities, right? Because before `forward()`, everything in the compute graph is symbolic(?). 

I don't know when a compute graph is destroyed, but if a compute graph can dynamically be a subgraph of a larger one, then it seems possible to have something like:
```rust
let x = neuronika::rand((1,)).requires_grad();
let y = x * 2;
let y_val = y.forward(); // evaluate a value from a subgraph
let z = if y_val > 1.0{
    y * 2
}else{
    y * 3
}; // add more compute nodes based on y_val
z.forward(); // evaluate a whole graph. Probably with redundant computations on the above subgraph? Or is there any cache?
z.backward();
```

**Comment by frjnn** [comment]
Hi, your example would be:

```rust
let x = neuronika::rand(1).requires_grad();
let y = x * 2;

y.forward(); // evaluate up until y, computations are cached

let z = if y.data()[0] > 1.0{
    y * 2
} else {
    y * 3
}; // add more nodes based on y.data()

z.forward(); // evaluate the subgraph from y to z
z.backward(1.0);
```

Yes, subgraph are cached and will not be recomputed.

**Comment by frjnn** [comment]
If you want to free the graph after back-propagating just use a scope:

```rust
let x = neuronika::rand(1).requires_grad();
{
    let y = x * 2;

    y.forward(); // evaluate up until y, computations are cached

    let z = if y.data()[0] > 1.0{
        y * 2
    } else {
        y * 3
    }; // add more nodes based on y.data()

    z.forward(); // evaluate the subgraph from y to z
    z.backward(1.0);
}
```

**Comment by ifsheldon** [comment]
Sorry if this becomes more like a Q&A. I am currently working on a differentiable physical simulator and thus searching for a good framework to work with. (Unfortunately PyTorch on Python is passed because of GIL)

My algorithm is something like the below pseudo-code:
```
def process_a(a):
    b = algorithm1(a_i)  # pure differentiable math function
    if b is not okay:
        return a_i
    c = algorithm2(b) # pure differentiable math function
    if c is not okay:
        return b
    return algorithm3(c) # pure differentiable math function

result_array = Array()
parameter = rand().needs_grad()

for i in range(100):  # can be a parallel for-loop without data races
    a = random_values(shape=(10, 10)) * parameter
    results = parallel_mapping(process_a, a)
    collect results into result_array

calc loss from result_array and backward.
```
The compute graph of my program will start at one node and end with one node, and there are a lot of parallel paths between the start and the end, whose length depends on intermediate values. I wish these parallel paths could be calculated in parallel. It's a bit complex and very dynamic, so I wonder if `neuronika` can handle this.

**Comment by frjnn** [comment]
Don't worry, you are more than welcome to ask any question. Your loop is equivalent to:

```rust
let x = neuronika::ones(1).requires_grad();
let y = x.clone() * 2. + 1.;

y.forward();

while y.data()[0] < 10. {
    let k = y.clone() * 2.;
    k.forward();
        
    y.data_mut().assign(&k.data());
    k.backward(1.0);
}
```

It's ugly and I know it, but the eager mode you've mentioned will not save the day. What you really need here is a dynamically typed variable, which we don't yet support, since the type of `y` changes at every iteration.

Concerning you algorithm, at first glance, what you are doing in that for loop is perfectly fine and you should not encounter the aforementioned issue.

**Comment by frjnn** [comment]
Hi, good news, I implemented such variables and now you are able to do such things pretty easily:

```rust
let mut x = neuronika::ones(()).requires_grad().into_dyn();
let y = x.clone();

for _ in 0..5 {
    x = (x.clone() * 4.).into_dyn();
}

x.forward();
x.backward(1.);
```

Do mind that this is not yet on main, but it will soon be.

> To me, the APIs are great, despite writing something like the loop is a bit confusing. And the only confusing part is k.backward(), which has a very different meaning from PyTorch's I think. In PyTorch, .backward() will accumulate gradients, but here it seems to just construct (add one compute node to) a backward compute graph(if you would).

neuronika's backward differentiates the graph and accumulates the gradients in the leaves, it could be that I don't properly understand your doubts.

Furthermore, you should be able to compute paths in parallel (from a safety point of view) as long as they are separate branches. Differentiating the graph is not thread safe currently, since the gradient is computed by summing signals coming from different parallel branches, but it could given a thread-safe variable, which can be done.

An eager evaluation would be straightforward to implement, it can be done by creating the variable and subsequently calling `forward()` on it. As you said, having direct control on `forward()` and `backward()` opens the possibility for further optimisations and the computational graph can be recycled.

**Comment by ifsheldon** [comment]
> neuronika's backward differentiates the graph and accumulates the gradients in the leaves, it could be that I don't properly understand your doubts.

I ran your code, this one
```rust
let x = neuronika::ones(1).requires_grad();
let y = x.clone() * 2. + 1.;

y.forward();

while y.data()[0] < 10. {
    let k = y.clone() * 2.;
    k.forward();
        
    y.data_mut().assign(&k.data());
    k.backward(1.0);
}
```
then the gradient of x I got is `8`. Mathematically, this makes sense because the computations is `(x*2+1)*2*2`. However, the confusing part is `k.backward()`. If I understand "accumulates the gradients in the leaves" correctly, then such code is equivalent to the below PyTorch code.
```python
x = torch.ones(1, requires_grad=True)
y = x * 2 + 1
while y.item() < 10.0:
    y = y * 2
    y.backward(retain_graph=True)

print(x.grad)
```
Then what you see will be `12` which is `4+8` because in the first iteration's backward, `x` gets a gradient of `4` and then in the second iteration, `x` gets another `8`.

> An eager evaluation would be straightforward to implement, it can be done by creating the variable and subsequently calling forward() on it. 

I will suggest a lazy and implicitly edge mode --only when data access operations like `.data()` are called, should `.forward` be called implicitly. In this way, we can have lazy subgraphs, which enables some optimizations, while we still have PyTorch-like convenience. But this is my suggestion, whether to take it is totally up to you.

**Comment by frjnn** [comment]
> Then what you see will be 12 which is 4+8 because in the first iteration's backward, x gets a gradient of 4 and then in the second iteration, x gets another 8.

Yes, in your case `y` gets updated, I was referring to your second snippet. Anyways, you should now be able to write equivalent things in neuronika. 

I will experiment with your suggestion, but as of now I think I should prioritize other things. By the way, having people using the project greatly helps me, thank you.

**Comment by ifsheldon** [comment]
> I was referring to your second snippet, it was not really equivalent but it gave the same result. Anyways, you should now be able to write properly equivalent things in neuronika

Yeah, I know your code snippet is equivalent to my second code snippet. What I meant is that `k.backward(1.0)` is conceptually confusing for me, because at the first sight, I have an impression that your code snippet is equivalent to 
```python
# your code snippet may be conceptually confused with this
# because in the loop, there is a line: k.backward(1.0)
x = torch.ones(1, requires_grad=True)
y = x * 2 + 1
while y.item() < 10.0:
    y = y * 2
    y.backward(retain_graph=True)

print(x.grad)
```
not this
```python
# equivalent to my second code snippet and your code snippet
x = torch.ones(1, requires_grad=True)
y = x * 2 + 1
while y.item() < 10.0:
    y = y * 2

y.backward()
print(x.grad)
```
By "confusing", I meant `k.backward(1.0)` in the Rust loop has a different meaning and functionality from `y.backward(retain_graph=True)` in the Python loop. So, probably you need to clarify a bit on this in case of that people from PyTorch have a wrong interpretation on your code snippet from the first sight.

**Comment by ifsheldon** [comment]
OK, got it. I just tried dynamic typing, and it seems dynamic typing does not broadcast to computed results / not have stickiness.
```rust
fn dyn_loop_wrong(){
    // compile error
    let x = neuronika::ones(1).requires_grad().into_dyn();
    let mut y = x.clone() * 2. + 1.;
    y.forward();
    while y.data()[0] < 10. {
        y = y * 2.;
        y.forward();
    }
    y.backward(1.0);
    println!("{:?}", x.grad())
}

fn dyn_loop(){
    // OK
    let x = neuronika::ones(1).requires_grad();
    let mut y = (x.clone() * 2. + 1.).into_dyn();
    y.forward();
    while y.data()[0] < 10. {
        y = (y * 2.).into_dyn();
        y.forward();
    }
    y.backward(1.0);
    println!("{:?}", x.grad())
}
```
Although writing `into_dyn()` in this way is a bit tedious, I think it makes sense in the context of "the lazy and implicitly eager mode".
